include/mbedtls/ssl.h

```diff
@@ -525,6 +526,160 @@ typedef void mbedtls_ssl_set_timer_t( void * ctx,
+
+/**
+ * \brief           Callback type: start external decryption operation
+ *
+ *                  Callback to start a decryption operation using an
+ *                  external processor. The parameter \c cert contains
+ *                  the public key; it is up to the callback function to
+ *                  look up the associated private key or a handle to the
+ *                  private key.
+ *
+ *                  This function must start the decryption operation.
+ *                  It is expected to be non-blocking, i.e. typically
+ *                  this function sends or enqueues a request and does
+ *                  not wait for the operation to complete.
+ *
+ *                  The parameters \c connection_ctx and \c cert are
+ *                  guaranteed to remain valid as long as the SSL
+ *                  configuration remains valid. On the other hand, this
+ *                  function must save the contents of \c hash, as the
+ *                  \c hash buffer is no longer valid when this function
+ *                  returns.
+ *
+ * \param connection_ctx  Pointer to the connection context set in the
+ *                        SSL configuration
+ * \param p_operation_ctx On output, pointer to the operation context.
+ *                        This pointer will be passed later to the resume
+ *                        or detach function. The value is only used if
+ *                        an operation is started, i.e. if this callback
+ *                        returns 0 or \c MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS.
+ * \param cert            Certificate containing the public key
+ * \param input           Buffer containing the input ciphertext. This buffer
+ *                        is no longer valid when the function returns.
+ * \param input_len       Size of the \c input buffer in bytes
+ *
+ * \return          - 0 if the SSL stack should call the resume callback
+ *                    immediately. The resume function may provide the
+ *                    or may itself return
+ *                    \c MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS.
+ *                  - \c MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS if the SSL stack
+ *                    should return immediately without calling the resume
+ *                    callback.
+ *                  - \c MBEDTLS_ERR_SSL_HW_ACCEL_FALLTHROUGH if the external
+ *                    processor does not support this key. The SSL stack will
+ *                    use the associated private key object instead.
+ *                  - Any other error is propagated up the call chain.
+ */
+typedef int mbedtls_ssl_async_decrypt_t( void *connection_ctx,
+                                         void **p_operation_ctx,
+                                         mbedtls_x509_crt *cert,
+                                         const unsigned char *input,
+                                         size_t input_len );
+#endif /* MBEDTLS_X509_CRT_PARSE_C */
```

Comment:
It's for potential reuse: it would be logical to extend this feature to pre-shared symmetric keys, and then you'd have a new callback to start using the PSK, and the same `resume` and `cancel` functions to finish the operation. That callback would have `_psk` in its name, so I think it's fine to keep `async_decrypt_t` here.